logging/log_processing/json_logging.py

```diff
@@ -0,0 +1,148 @@
+"""
+Enable logging (from Lambdas) in JSON format which makes post-processing much easier.
+
+Since we assume this will be used by Lambdas, we also add the request id in log lines.
+"""
+
+import json
+import logging
+import logging.config
+import time
+
+
+class ContextFilter(logging.Filter):
+    """
+    Logging Filter class that adds contextual information to log records.
+
+    We assume there will be only one instance of this filter for any runtime which
+    means that we will store some values with the class, not the instances.
+    """
+
+    _context = {
+        "aws_request_id": None,
+        "function_name": None,
+        "function_version": None,
+        "log_group_name": None,
+        "log_stream_name": None,
+    }
+
+    def filter(self, record):
+        """
+        Modify record in place for additional fields, then return True to continue processing.
+        """
+        for field, value in self._context.items():
+            if value is not None:
+                setattr(record, field, value)
+        return True
+
+    @classmethod
+    def update_context(cls, **kwargs):
+        """
+        Update any of the fields stored in the (global) context filter.
+
+        Note that trying to set a field that's not been defined raises a ValueError.
+        """
+        for field, value in kwargs.items():
+            if field in cls._context:
+                cls._context[field] = value
+            else:
+                raise ValueError(f"unexpected field: '{field}'")
+
+
+class JsonFormatter(logging.Formatter):
+    """
+    Format the message to be easily reverted into an object by using JSON format.
+
+    Notes:
+        * The "format" is ignored since we convert based on available info.
+        * The timestamps are in UTC.
+    """
+
+    # This format is compatible with "strict_date_time" in Elasticsearch (yyyy-MM-dd'T'HH:mm:ss.SSSZZ).
+    converter = time.gmtime
+    default_time_format = "%Y-%m-%dT%H:%M:%SZ"
+    default_msec_format = "%.19s.%03dZ"
+
+    attribute_mapping = {
+        # LogRecord attributes for which we want new names:
+        "filename": "source:filename",
```

Comment:
Is there a reason why this uses a colon separator, whereas everything else uses a dot?

Reply:
Fixed in a later PR.